Fix bug of link query order on markdown render

lunny opened the pull request:
Should fix #14090

We temporarily use a replace on go.mod before https://github.com/microcosm-cc/bluemonday/pull/105 merged.

6543 commented:
@lunny need to change tests:
```
--- FAIL: TestRender_links (0.02s)
    html_test.go:89: 
        	Error Trace:	html_test.go:89
        	            				html_test.go:125
        	Error:      	Not equal: 
        	            	expected: "<p><a href=\"https://github.com/go-gitea/gitea/?p=aaa%2Fbbb.html#ccc-ddd\" rel=\"nofollow\">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>"
        	            	actual  : "<p><a href=\"https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd\" rel=\"nofollow\">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-<p><a href="https://github.com/go-gitea/gitea/?p=aaa%2Fbbb.html#ccc-ddd" rel="nofollow">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>
        	            	+<p><a href="https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd" rel="nofollow">https://github.com/go-gitea/gitea/?p=aaa/bbb.html#ccc-ddd</a></p>
        	Test:       	TestRender_links
    html_test.go:89: 
        	Error Trace:	html_test.go:89
        	            				html_test.go:143
        	Error:      	Not equal: 
        	            	expected: "<p><a href=\"magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	actual  : "<p><a href=\"magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-<p><a href="magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	            	+<p><a href="magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	Test:       	TestRender_links
FAIL
```

zeripath commented:
no - don't change the tests - fix the query parameter it's not escaping properly.

lunny commented:
The first test should be right and I changed the bluemonday codes.

I also changed the second test on this PR because it should be a wrong test.

```go
html_test.go:143
        	Error:      	Not equal: 
        	            	expected: "<p><a href=\"magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	actual  : "<p><a href=\"magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download\" rel=\"nofollow\">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-<p><a href="magnet:?dn=download&xt=urn%3Abtih%3A5dee65101db281ac9c46344cd6b175cdcadabcde" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	            	+<p><a href="magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&dn=download" rel="nofollow">magnet:?xt=urn:btih:5dee65101db281ac9c46344cd6b175cdcadabcde&amp;dn=download</a></p>
        	Test:       	TestRender_links
```

It's obvious, the query order changed but should not.

zeripath commented on vendor/github.com/microcosm-cc/bluemonday/sanitize.go:
```diff
@@ -122,22 +122,79 @@ func escapeUrlComponent(val string) string {
-			sanitizedQueryValues.Add(sk, sv)
-		}
+	for i, query := range queryValues {
+		queryValues[i].Key = html.EscapeString(query.Key)
```
I don't understand this. What's the aim of this? url.QueryEscape should do for both the key and value.

html.EscapeString will still need to be run on the whole url.

lunny replied:
Hm, I haven't look into this, but just translate from the old code of bluemonday. See https://github.com/microcosm-cc/bluemonday/pull/105/files
